File: composeApp/src/androidMain/kotlin/com/tayler/pizzzaapp/ui/AppViewModel.kt
package com.tayler.pizzzaapp.ui

import android.util.Log
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.tayler.pizzzaapp.entity.ParentOrderModel
import com.tayler.pizzzaapp.entity.ProductModel
import com.tayler.pizzzaapp.ui.base.BaseViewModel
import com.tayler.pizzzaapp.usecases.DataUseCase
import com.tayler.pizzzaapp.utils.DispatcherProvider

import kotlinx.coroutines.withContext

data class OrderUiState(
    val orders: List<ParentOrderModel> = emptyList(),
    val filteredOrders: List<ParentOrderModel> = emptyList(),
    val selectedFilter: String = "TODOS",
    val countConfirmado: Int = 0,
    val countListo: Int = 0,
    val selectedOrder: ParentOrderModel? = null,
    val products: List<ProductModel> = emptyList()
)

class AppViewModel(
    private val dataUseCase: DataUseCase,
    private val dispatchers: DispatcherProvider
) : BaseViewModel(dispatchers) {

    var orderUiState by mutableStateOf(OrderUiState())
        private set

    fun getGeneralOrderList() {
        Log.d("AppViewModel", "getGeneralOrderList: Iniciando ejecución")
        execute {
            try {
                val response = dataUseCase.loadParentOrder()
                updateStateWithOrders(response)
            } catch (e: Exception) {
                Log.e("AppViewModel", "Error en getGeneralOrderList: ${e.message}", e)
                throw e
            }
        }
    }

    fun refresh() {
        Log.d("AppViewModel", "refresh: Forzando refresco")
        execute {
            try {
                val response = dataUseCase.loadParentOrder(forceRefresh = true)
                updateStateWithOrders(response)
            } catch (e: Exception) {
                Log.e("AppViewModel", "Error en refresh: ${e.message}", e)
                throw e
            }
        }
    }

    private fun updateStateWithOrders(orders: List<ParentOrderModel>) {
        val sortedOrders = orders.sortedBy {
            when (it.state.trim().uppercase()) {
                "CONFIRMADO" -> 1
                "LISTO" -> 2
                else -> 3
            }
        }

        val countConfirmado = orders.count { it.state.trim().uppercase() == "CONFIRMADO" }
        val countListo = orders.count { it.state.trim().uppercase() == "LISTO" }

        orderUiState = orderUiState.copy(
            orders = sortedOrders,
            filteredOrders = sortedOrders, // Mostramos todos por defecto ya que no hay filtros
            countConfirmado = countConfirmado,
            countListo = countListo
        )
    }

    fun applyFilter(filter: String) {
        val filtered = if (filter == "TODOS") {
            orderUiState.orders
        } else {
            orderUiState.orders.filter { it.state.trim().uppercase() == filter.uppercase() }
        }
        orderUiState = orderUiState.copy(
            filteredOrders = filtered,
            selectedFilter = filter
        )
    }

    fun updateOrderState(order: ParentOrderModel, newState: String) {
        if (order.state.trim().uppercase() == newState.uppercase()) return

        // 1. Guardar estado previo para Reversión (Rollback) en caso de error
        val previousState = orderUiState

        // 2. Actualización Optimista: Actualizamos la UI inmediatamente
        Log.d("AppViewModel", "updateOrderState: Actualización optimista de ${order.uid} a $newState")
        val updatedOrders = orderUiState.orders.map { 
            if (it.uid == order.uid) it.copy(state = newState) else it 
        }
        updateStateWithOrders(updatedOrders)

        // 3. Sincronización en segundo plano
        // Usamos loading = false para que no aparezca el progreso global y la app se sienta "rápida"
        execute(loading = false) {
            try {
                dataUseCase.updateOrder(order.copy(state = newState))
                Log.d("AppViewModel", "updateOrderState: Sincronización exitosa con servidor")
            } catch (e: Exception) {
                // 4. Rollback: Si falla el servidor, devolvemos la UI a su estado anterior
                Log.e("AppViewModel", "updateOrderState: Error al sincronizar. Revirtiendo UI.", e)
                orderUiState = previousState
                throw e // Permitimos que BaseViewModel muestre el diálogo de error
            }
        }
    }

    fun avanzarEstado(order: ParentOrderModel) {
        val currentState = order.state.trim().uppercase()
        val nextState = when (currentState) {
            "CONFIRMADO" -> "LISTO"
            "LISTO" -> "ENTREGADO" // O el estado final que manejes
            else -> "CONFIRMADO"
        }
        updateOrderState(order, nextState)
    }

    fun getProductsList() {
        val hasData = orderUiState.products.isNotEmpty()
        // Si ya hay datos, cargamos en segundo plano para no bloquear
        execute(loading = !hasData) {
            try {
                val response = dataUseCase.getProducts()
                withContext(dispatchers.main) {
                    orderUiState = orderUiState.copy(products = response)
                }
            } catch (e: Exception) {
                Log.e("AppViewModel", "Error en getProductsList: ${e.message}", e)
                throw e
            }
        }
    }

    fun selectOrder(order: ParentOrderModel?) {
        orderUiState = orderUiState.copy(selectedOrder = order)
    }
}

File: iosApp/iosApp/AppViewModel.swift
import Foundation
import Shared
import Combine

class AppViewModel: ObservableObject {
    @Published var pizzaProducts: [ProductModel] = []
    @Published var extraProducts: [ProductModel] = []
    @Published var deliveryProducts: [ProductModel] = []
    @Published var orders: [ParentOrderModel] = []
    @Published var selectedProduct: ProductModel? = nil
    @Published var selectedOrder: ParentOrderModel? = nil
    @Published var isLoading: Bool = false
    @Published var ordersLoaded: Bool = false
    
    private let dataUseCase = KoinHelper.shared.getDataUseCase()
    
    func syncProducts(onComplete: @escaping (Bool) -> Void) {
        isLoading = true
        dataUseCase.syncProducts { [weak self] _, error in
            if error != nil {
                DispatchQueue.main.async {
                    self?.isLoading = false
                    onComplete(false)
                }
                return
            }
            
            self?.getProductsList { success in
                onComplete(success)
            }
        }
    }
    
    func getProductsList(onComplete: ((Bool) -> Void)? = nil) {
        dataUseCase.getProducts { [weak self] response, error in
            DispatchQueue.main.async {
                self?.isLoading = false
                if let products = response {
                    self?.pizzaProducts = products.filter { $0.type == "1" }
                    self?.extraProducts = products.filter { $0.type == "2" || $0.type == "3" }
                    self?.deliveryProducts = products.filter { $0.type == "4" }
                    onComplete?(true)
                } else {
                    onComplete?(false)
                }
            }
        }
    }
    
    func getGeneralOrderList(forceLoading: Bool = false) {
        if ordersLoaded && !forceLoading { return }
        
        isLoading = forceLoading
        dataUseCase.getUserLocal { [weak self] user, error in
            guard let uid = user?.uid else {
                DispatchQueue.main.async {
                    self?.isLoading = false
                    self?.orders = []
                }
                return
            }
            
            self?.dataUseCase.loadParentOrder(userId: uid) { response, error in
                DispatchQueue.main.async {
                    self?.isLoading = false
                    if let orders = response {
                        self?.updateStateWithOrders(orders: orders)
                    }
                }
            }
        }
    }
    
    private func updateStateWithOrders(orders: [ParentOrderModel]) {
        self.orders = orders.sorted { o1, o2 in
            func priority(_ s: String) -> Int {
                switch s.trimmingCharacters(in: .whitespacesAndNewlines).uppercased() {
                case "CONFIRMADO": return 1
                case "LISTO": return 2
                default: return 3
                }
            }
            return priority(o1.state) < priority(o2.state)
        }
        self.ordersLoaded = true
    }
    
    func selectProduct(_ product: ProductModel?) {
        self.selectedProduct = product
    }
    
    func selectOrder(_ order: ParentOrderModel?) {
        self.selectedOrder = order
    }
}
